### DISCUSSION TITLE
"native,Version=v0.0" Visual Studio 2017 project C++/CLI

### DISCUSSION THREAD

**Discussion by PRZEMPM** [NuGet/Home#12671] [General]
### The package "Peak.PCANBasic.NET 4.7.0.793" could not be installed. You are trying to install this package in a project that targets "native,Version=v0.0", but the package does not contain any assembly references or content files that match this structure. Contact the package author for more information.

I'm having trouble downloading any package using NuGet. The project C++/CLI compiles and runs, but as a GUI it is useless because I can't add the library. What should I do step by step in the project C++/CLI to solve this problem?

Helpful links:
https://docs.peak-system.com/API/PCAN-Basic.Net/html/3f24583f-9271-4de5-80d2-6dc999d58b10.htm#Answer7

**Comment by zivkan** [comment]
@pwr230047 C++/CLI projects only support .NET NuGet packages from recent versions of Visual Studio 2022 (from memory 17.4 or above. I expect our release notes should mention the specific version it was added). Prior to this, for example in VS 2017, C++/CLI projects can only use "native" packages (same as non-CLI C++ project)

**Reply by PRZEMPM** [reply]
I am using Visual Studio 2017 for the first time if I have versions 15.9.55, and I cannot have Visual Studio 2022 because I do not have a license.
What solution will you propose?

**Reply by zivkan** [reply]
C++/CLI only works on Windows, so anyone who wants to interop between C++ and .NET code on Mac, Linux, Android, iOS, etc, they all have create a native library with C bindings, and then P/Invoke it from .NET code. You could do the same.

Is C++/CLI able to use ProjectReferences to managed languages, like C# projects?  If so, you could create C# Class Library that provides "adapters" for any NuGet packages you want to use. Alternatively, make your C++/CLI code a class library, that only provides .NET APIs for native code, and then do all the interop between your C++/CLI class library and NuGet packages in a managed .NET language. Although I'm not sure if project references between C++ and C# projects are something that was also broken before VS2022.

**Reply by PRZEMPM** [reply]
I am going to write applications ONLY for Windows 10.
As I mentioned, I don't know much about Visual Studio, so everything you wrote is probably true, but you wrote too general for me and in Microsoft language, which I don't understand.
I have packages installed ONLY for C++/CLI. I didn't install the C# set, because what's the point if I only use C++??
1. I don't understand how to link the created native library and what library is it?
2. What are "adapters"?
3. How to set kd C++/CLI as class library?

**Reply by PRZEMPM** [reply]
**Solution:**
[https://itecnote.com/tecnote/c-how-to-make-the-managed-nuget-package-support-c-cli-projects/](url)

**Reply by zivkan** [reply]
> As I mentioned, I don't know much about Visual Studio

Did you intend to create a .NET project (C++/CLI), or were you trying to create a "normal" C++ project?  If you're trying to create a native C++ project, then C++/CLI is not what you're looking for, you should create a new C++ console app (not C++/CLI) from the new project wizard, and you can ignore everything I write below. But in that case it's expected that the package won't install into the C++ project, because the package is for .NET.

> I have packages installed ONLY for C++/CLI. I didn't install the C# set, because what's the point if I only use C++??

Do you mean workloads, in the Visual Studio Installer? If you did mean package (NuGet package?), then I don't understand the question. Otherwise, the point of installing the .NET workload is because you're trying to integrate a .NET package (Peak.PCANBasic.NET) with C++ code. VS2017, even VS2019, does not support C++/CLI projects referencing .NET NuGet packages. Given these technical constraints you either need to find a C/C++ library to use instead, or use a C# project to "connect" the package to the rest of your code. If you don't want to install the .NET workload and you can't upgrade to Visual Studio 2022, then using the .NET NuGet package is just not possible (via NuGet's integration in Visual Studio).

You might be able to avoid using NuGet's Visual Studio integration by manually downloading the package, extract it somewhere (it's just a zip file renamed), and then reference the dll directly in the project. But then you either need to commit the dll into source control, or find another solution that doesn't involve NuGet and the C++/CLI project system being integrated.

> I don't understand how to link the created native library and what library is it?

I don't understand your question. If your question is about calling native libraries from C# (of VB.NET, or F#), then it's done through something .NET calls Platform Invoke (P/Invoke): https://learn.microsoft.com/en-us/dotnet/standard/native-interop/pinvoke

> What are "adapters"?

Adapters are a commonly used software design pattern: https://en.wikipedia.org/wiki/Adapter_pattern

> How to set kd C++/CLI as class library?

I don't know. NuGet is written entirely in c#, so our team doesn't have expertise with C++/CLI projects. Visual Studio's C++ teams are not active on GitHub, so we can't help you here. You could try asking on https://learn.microsoft.com/answers, or stack overflow. There are probably also Discord servers for people learning .NET. I don't know how active reddit.com/r/dotnet is any longer, but searching "reddit c++/cli" shows many reddit posts across several subreddits.

**Comment by warlocklp** [comment]
Hi bro, So what can I do for this? URL is invalid.
